### DISCUSSION TITLE
Make `Mapped` covariant

### DISCUSSION THREAD

**Discussion by RomeoDespres** [sqlalchemy/sqlalchemy#10287] [Ideas]
Hi friends, I've noticed that `Mapped` is currently invariant with its type parameter. What would you think of making it covariant?

I came accross this when trying to build protocols to represent some of my data tables. I was trying to write a protocol representing a table, and then another protocol representing another table that has the first protocol as a relationship. Complete example below.

The aim is to be able to write general helper functions (like `get_parent_name`) that operate on table protocols instead of actual tables.

```python
from typing import Protocol

from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship


# Protocols


class ParentProtocol(Protocol):
    name: Mapped[str]


class ChildProtocol(Protocol):
    # Read-only for simplicity, mutable protocol members are complicated,
    # see https://mypy.readthedocs.io/en/latest/common_issues.html#covariant-subtyping-of-mutable-protocol-members-is-rejected
    @property
    def parent(self) -> Mapped[ParentProtocol]:
        ...


def get_parent_name(child: ChildProtocol) -> str:
    return child.parent.name


# Implementations


class Base(DeclarativeBase):
    pass


class Parent(Base):
    __tablename__ = "parent"

    name: Mapped[str] = mapped_column(primary_key=True)


class Child(Base):
    __tablename__ = "child"

    name: Mapped[str] = mapped_column(primary_key=True)
    parent_name: Mapped[str] = mapped_column(ForeignKey(Parent.name))

    parent: Mapped[Parent] = relationship()


assert get_parent_name(Child(parent=Parent(name="foo"))) == "foo"
```

The code above runs fine but does not type-check because of the invariance of `Mapped`:

```
test.py:48: error: Argument 1 to "get_parent_name" has incompatible type "Child"; expected "ChildProtocol"  [arg-type]
test.py:48: note: Following member(s) of "Child" have conflicts:
test.py:48: note:     parent: expected "Mapped[ParentProtocol]", got "Mapped[Parent]"
Found 1 error in 1 file (checked 1 source file)
```

With a few changes in sqlalchemy source code I've managed to make `Mapped` covariant and mypy run fine on the codebase + this example. Not sure though if there are some deeper implications. If you think it's interesting I can submit a PR and keep on discussing there

**Comment by zzzeek** [comment]
hi -

seems like this would be fine.   For typing related PRs we would add your example above to likely a new file `test/typing/plain_files/orm/mapped_covariant.py` - we run mypy on all those files so that would be the passing case here.  then everything else would we assume keep passing as well.     I don't see a reason this would create a problem, so sure let's try it out w a PR including that test file.

**Reply by RomeoDespres** [reply]
thanks! will post it this week then!

**Reply by jace** [reply]
Did this happen? I've run into a similar problem attempting to write a mixin class whose methods use a Protocol to specify subclasses. Something like this:

```
from typing import Protocol
from uuid import UUID, uuid4

import sqlalchemy as sa
from sqlalchemy.orm import DeclarativeBase, Mapped, declared_attr


class ModelBase(DeclarativeBase):
    pass


class CompareProtocol(Protocol):
    id: Mapped[int | UUID]  # Or `Mapped[int] | Mapped[UUID]`


class CompareMixin:
    def compare(self: CompareProtocol, other: CompareProtocol) -> bool:
        return self.id == other.id  # Simplified example


class IntIdMixin:
    @declared_attr
    def id(cls) -> Mapped[int]:
        return sa.orm.mapped_column(sa.Integer, primary_key=True)


class UuidIdMixin:
    @declared_attr
    def id(cls) -> Mapped[UUID]:
        return sa.orm.mapped_column(sa.UUID, primary_key=True, default=uuid4)


class MyModel(CompareMixin, IntIdMixin, ModelBase):
    pass


m1 = MyModel()
m2 = MyModel()
m1.compare(m2)  # Mypy reports two errors here
```
Errors:

1. Invalid self argument "MyModel" to attribute function "compare" with type "Callable[[CompareProtocol, CompareProtocol], bool]"  [misc]
2. Argument 1 to "compare" of "CompareMixin" has incompatible type "MyModel"; expected "CompareProtocol"  [arg-type]
    * Following member(s) of "MyModel" have conflicts:
    * id: expected "Mapped[int | UUID]", got "declared_attr[int]"
    * Protocol member CompareProtocol.id expected settable variable, got read-only attribute

The key issue seems to be that any specification of `Mapped[…]` in the protocol must be an exact match in the implementing class — with no scope for flexibility. Using `declared_attr` in the protocol isn't a workaround either — it's interpreted as a read-only attribute and the mixin can't attempt to set it.

**Reply by CaselIT** [reply]
Was released in 2.0.21 https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-4912ee2ba5731476579dbb664755253a

**Reply by jace** [reply]
Looks like my use case is still not covered (as of 2.0.23) because `Mapped` in the Protocol doesn't match `declared_attr -> Mapped` in the implementation.

**Reply by CaselIT** [reply]
Can you propose a PR with the fix?

**Reply by jace** [reply]
Type variancy is still a bit fuzzy to me, but I'll try.

**Reply by CaselIT** [reply]
let's open a new issue for this so it gets tracked

**Reply by CaselIT** [reply]
I've created #10673
